### ISSUE TITLE
Flag `env!` macro usage in build scripts

### ISSUE THREAD

**Issue by d-e-s-o** [rust-lang/rust-clippy#15872]
### What it does

I am not sure if lints can or should make assumptions about the file they are defined in, but it would be good to flag usage of the [`env!`](https://doc.rust-lang.org/std/macro.env.html) (and `option_env!`) macro in build scripts. In all likelihood this is wrong. It certainly is wrong for `CARGO_*` variables [[0]](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts).

### Advantage

It is easy to miss this wrong usage, as it may not break immediately.

### Drawbacks

None that I can think of.

### Example

```rust
env!("CARGO_PKG_VERSION")
```

Could be written as:

```rust
std::env::var("CARGO_PKG_VERSION").expect("CARGO_PKG_VERSION variable not set")
```

### Comparison with existing lints

Not aware of existing lints doing something similar.
